controller/src/main/java/io/pravega/controller/store/stream/ZKScope.java

```diff
@@ -37,16 +72,207 @@ public String getName() {
 
     @Override
     public CompletableFuture<Void> deleteScope() {
-        return store.deleteNode(scopePath);
+        return store.deleteNode(scopePath)
+                    .thenCompose(v -> Futures.exceptionallyExpecting(store.deleteTree(counterPath), DATA_NOT_FOUND_PREDICATE, null))
+                    .thenCompose(v -> Futures.exceptionallyExpecting(store.deleteTree(streamsInScopePath), DATA_NOT_FOUND_PREDICATE, null));
+    }
+
+    CompletableFuture<Void> addStreamToScope(String name, int streamPosition) {
+        // break the path from stream position into three parts --> 2, 4, 4
+        String path = getPathForStreamPosition(name, streamPosition);
+        return Futures.toVoid(store.createZNodeIfNotExist(path));
+    }
+
+    CompletableFuture<Void> removeStreamFromScope(String name, int streamPosition) {
+        String path = getPathForStreamPosition(name, streamPosition);
+
+        return Futures.toVoid(store.deletePath(path, true));
+    }
+
+    private String getPathForStreamPosition(String name, int streamPosition) {
+        Preconditions.checkArgument(streamPosition >= 0);
+        Token token = new Token(streamPosition);
+
+        // root path represents most significant 2 digits of the position.
+        String root = ZKPaths.makePath(streamsInScopePath, token.getMsb().toString());
+        // container path represents most significant 3rd to 6th digits of the position.
+        String container = ZKPaths.makePath(root, token.getMiddle().toString());
+        // the path is composed of stream name followed by last 4 chars that represent least significant 4 digits of the position.
+        return ZKPaths.makePath(container, String.format("%s%04d", name, token.getLsb()));
+    }
+
+    private String getStreamName(String x) {
+        return x.substring(0, x.length() - 4);
+    }
+
+    private int getPosition(String x) {
+        return Integer.parseInt(x.substring(x.length() - 4, x.length()));
     }
 
     @Override
     public CompletableFuture<List<String>> listStreamsInScope() {
         return store.getChildren(scopePath);
     }
 
+    @Override
+    public CompletableFuture<Pair<List<String>, String>> listStreamsInScope(int limit, String continuationToken, Executor executor) {
```

Comment:
I haven't reviewed this logic carefully, but my suggestion is that we add more comments here to make it easier to follow what it is trying to do. In particular, I'd like to understand how this is paginating.